### ISSUE TITLE
Large-model memory ceiling — 4 engine capabilities needed (Float16/BFloat16, weight streaming, quantization, GPU offload)

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#276]
## Summary

Profiling the paper-faithful PaLM-E 562B config in `AiDotNet` (Driess et al. 2023, see ooples/AiDotNet#1222) revealed that the engine cannot fit the model's ~140 GB of weights at `double` precision. Forward pass OOMs at 255 s on first weight materialization. The signal is real — that test exists specifically to catch perf bugs at scale — and the fixes all sit in this repo.

This is a rollup tracking four independent capabilities. Each can ship separately; AiDotNet picks them up transparently because every consumer already routes weights through `Tensor<T>` and `TensorAllocator`.

---

## Repro / context

```
Architecture:    1 ms
PaLME ctor:      37 ms
  Layers.Count = 678 (48 vision MHA + 64 decoder MHA + LayerNorms + FFNs)
  GC alloc through ctor: 2.0 MB    <- lazy weights, fine
ParameterCount:  0 ms  (saturated at int.MaxValue, real count = 17,561,094,400)
Input prep:      16 ms
Predict FAILED at 255223 ms: OutOfMemoryException
```

Memory math at `T = double`:
- Vision MHA: 4 x 1408^2 x 8 B = 63 MB each x 48 = 3 GB
- Decoder MHA: 4 x 8192^2 x 8 B = 8.6 GB each x 64 = **549 GB**
- FFN expansions: same order again
- Total live weights ~ 140 GB compute-side, > 200 GB worst case

PaLME is the canary, but the same ceiling applies to any large transformer (Stable Diffusion XL, Whisper-large, LLaMA-class) once they grow past ~5 B parameters in `double`.

---

### (1) Float16 / BFloat16 numeric types

**Problem.** AiDotNet defaults to `T = double` for tests and `T = float` for production. Modern transformer weights are stored at half precision (bfloat16 in PaLM-E, fp16 in Stable Diffusion, Float16 in Whisper). Storing ~17 B params at `double` is 8x the memory footprint of bfloat16.

**Scope of work in this repo.**
- Add `BFloat16` and `Float16` (IEEE 754-2008 binary16) value types under `AiDotNet.Tensors.Numerics`. .NET 5+ has `System.Half` — likely correct to wire that as `Float16` rather than reinvent it. BFloat16 has no BCL equivalent and must be implemented (~30 lines: `struct BFloat16 { ushort _bits; }` plus `IEquatable`, conversions to/from `float`).
- Implement `INumericOperations<BFloat16>` and `INumericOperations<Half>`. Most ops are `(BFloat16) => float => op => BFloat16` round-trips for correctness; the SIMD kernels are where the actual perf gain lives.
- Wire SIMD kernels for the hot paths: matmul, conv2d, layer-norm, gelu/silu/softmax, reduce-mean/sum. AVX-512 has `VCVTNE2PS2BF16` / `VCVTNEPS2BF16` for direct bf16 conversion; AVX2 falls back to bit-shift truncation (which is what bfloat16 round-to-nearest-even reduces to).
- Update the autodiff tape so backward kernels accept the new types. Mostly a question of letting the existing `INumericOperations<T>` indirection do its job — but bf16 has so little dynamic range that gradient accumulation needs to happen in `float` and downcast at the end (the standard "mixed precision" pattern, see Micikevicius et al. 2017).

**Validation.** Once shipped, AiDotNet can run `PaLME<BFloat16>` and the 549 GB decoder MHA shrinks to 137 GB. Still doesn't fit on commodity hardware, but it's the prerequisite for everything else and a free 4x win across the rest of the model zoo.

**Estimated complexity.** Medium. Type plumbing is mechanical; SIMD kernels are the real work. ~2 weeks of focused effort.

---

### (2) Weight streaming / on-demand materialization

**Problem.** Lazy layers defer weight allocation until first Forward, but on first Forward EVERY layer in the chain materializes its weights at the same time and they all stay live for the lifetime of the model. There's no chunking / streaming / offloading mechanism. For a 678-layer model that means 140 GB+ resident.

**Scope of work in this repo.**
- Extend `TensorAllocator` (or its successor) with a "streaming pool" mode. The pool tracks last-used time per registered weight tensor; when total resident bytes exceeds a budget, oldest unused weights are evicted to a backing store (memory-mapped file, NVMe, or compressed in-memory) and rehydrated on demand.
- Add a registration hook on `Tensor<T>` — when a layer constructs a weight tensor, it tags it with a `WeightLifetime.Streaming` flag and the allocator owns the pageability decision.
- The backward pass needs to ensure the weight is paged-in for the corresponding forward op's gradient computation. Easiest approach: prefetch the weight when the autodiff tape replays that op's backward — single-line addition at the dispatch point.
- Plumb a `StreamingPoolOptions` config (max resident bytes, eviction policy, backing store path) through `IEngine`.

**Validation.** With a 16 GB resident budget on a 64 GB box, 140 GB of weights stream through the working set. Per-step latency increases (NVMe page-in is ~3 GB/s sequential), but the model becomes runnable instead of OOMing. Exactly the trade-off DeepSpeed ZeRO-Offload makes — there's a published reference for the eviction policy.

**Estimated complexity.** Medium-large. The allocator side is straightforward; the autodiff prefetch coordination is where the bugs hide. Backward replay must not hold references to evicted tensors. ~3 weeks.

---

### (3) Quantized weight types (int8 / int4)

**Problem.** Even bf16 isn't enough for the largest models. Production deployments (LLaMA, Mistral, GPT-Q) use int8 or int4 weight quantization with float activations — gives a 4x / 8x memory reduction over bf16 and a measurable inference speedup because the dequantize-on-read fits in cache.

**Scope of work in this repo.**
- New tensor wrapper types: `QuantizedTensor<int8>` and `QuantizedTensor<int4>`. Each carries a `Tensor<byte>` payload, per-channel `scale` and `zero_point` vectors, and a quantization scheme enum (symmetric vs asymmetric). int4 packs two values per byte.
- Calibration API: `QuantizedTensor<int8>.FromFloat(Tensor<float>, QuantizationScheme)`. Reference impl is straightforward absmax-per-channel.
- Fused dequant-matmul kernels — these are where the perf win shows up. AVX2 has `VPMADDUBSW` (int8 x int8 -> int16 accumulate) plus a `VPMULPS` for the scale apply. Reference: `llama.cpp`'s `Q8_0` / `Q4_0` kernels.
- Autodiff: gradients accumulate in float, weights re-quantize at the end of each optimizer step. The QAT (Quantization-Aware Training) schedule is configurable per the paper.

**Validation.** PaLME at int8 = 17.5 GB of weights — fits on a workstation. int4 = 8.75 GB — fits in VRAM on a 12 GB consumer GPU.

**Estimated complexity.** Large. Calibration and the kernels are each multi-week work, but each ships independently (start with int8 inference-only, add training and int4 incrementally). ~6 weeks for full int8 + int4 round trip.

---

### (4) GPU offload with pinned host memory

**Problem.** The CUDA/OpenCL engine already exists and handles activations on-device, but weights still live in host memory and copy via PCIe per-op. For large models that don't fit in VRAM, the right pattern is unified memory or pinned-host buffers that the GPU can DMA directly without an explicit `cudaMemcpy`.

**Scope of work in this repo.**
- Extend the CUDA engine with a `PinnedHostAllocator` that wraps `cudaHostAlloc(cudaHostAllocPortable | cudaHostAllocMapped)`. Weights tagged `WeightLifetime.GpuOffload` allocate from this pool.
- The matmul / conv kernels read weights via mapped pointers — no explicit `cudaMemcpyAsync` per op. The CUDA driver overlaps the DMA with compute on the device side.
- For models that fit in unified-memory (Pascal+ GPUs, all consumer cards 2016+), use `cudaMallocManaged` instead — driver does the page migration automatically and there's no per-op staging cost.
- A `GpuOffloadOptions.PreferredScheme` config: explicit pinned vs managed vs automatic-best-fit based on `cudaDeviceProp::pageableMemoryAccess`.

**Validation.** PaLME at bf16 + int8 = 4-9 GB weights. Streams from host memory onto an 8 GB consumer GPU at PCIe 4.0 bandwidth (~30 GB/s, so the entire model traverses in ~300 ms — same order as the first MHA's compute, hides most of the cost).

**Estimated complexity.** Medium. CUDA-specific plumbing, plus an OpenCL equivalent (`clEnqueueMapBuffer` with `CL_MEM_ALLOC_HOST_PTR`) for AMD. ~3 weeks.

---

## Suggested sequencing

1. **Float16/BFloat16** first. Unblocks 4x memory reduction across every model, and is a prerequisite for (3) and (4) — quantization needs a float type to dequantize INTO; GPU offload pays back faster when you're moving fewer bytes per weight.
2. **GPU offload** second. Mostly self-contained CUDA work and immediately useful at workstation scale.
3. **Weight streaming** third. Decouples model size from physical RAM but adds complexity to the autodiff scheduler.
4. **Quantization** fourth. Largest scope, biggest payoff, and the existing int8 reference implementations (llama.cpp, GPTQ) make the algorithm work straightforward — it's the kernel work that's expensive.

## Cross-references

- AiDotNet canary that surfaced this: ooples/AiDotNet#1222
- Other Tensors-side blocker filed alongside: ooples/AiDotNet.Tensors#274 (PermuteBackward in-place-add contiguity)

**Comment by ooples** [comment]
Tracked in PR #277. All four sub-features ship together with cross-backend coverage (CUDA / HIP / Metal / OpenCL / Vulkan / WebGPU each get the offload-allocator implementation):

* (1) BFloat16 type + INumericOperations + MathHelper registry — same shape as the existing HalfOperations path.
* (2) StreamingTensorPool with LRU eviction to memory-mapped backing store.
* (3) QuantizedTensor<T> int8 / int4 with FromFloatInt8 / FromFloatInt4 calibration (GGUF Q8_0 / Q4_0 compatible).
* (4) IGpuOffloadAllocator with per-backend impls — pinned + managed schemes plumbed through every direct-GPU runtime.

Bonus infrastructure fix: NativeLibraryResolverRegistry consolidates 8 separate SetDllImportResolver calls into one shared dispatch chain. The .NET runtime only allows one resolver per assembly, so the prior 'last-cctor-wins' setup was a pre-existing latent bug that surfaced on hosts where multiple GPU probes loaded.

25 new integration tests covering the cross-backend matrix; full suite 4176/4176 passing on net10.0; net471 build clean across all 13 csproj.

**Comment by ooples** [comment]
All 7 audit gaps closed in PR #277 (commit 5306b14):

**Gap 1 — Tensor.Lifetime field + WeightRegistry routing**: Tensor<T> now carries Lifetime / StreamingPoolHandle / OffloadDevicePointer. WeightRegistry.RegisterWeight routes by lifetime to the right pool/allocator with per-element-type fast paths.

**Gap 2 — BFloat16 SIMD kernels**: BFloat16Kernels.cs ships AVX2/FMA paths for VectorAdd / VectorMultiply / Dot / ReduceSum / Matmul / GELU / SiLU / Softmax / LayerNorm operating directly on bf16 storage with float-accumulate. Exact bf16→float widening (upper-16-bits shift); RNE narrowing on writeback.

**Gap 3 — Streaming-pool autodiff prefetch**: StreamingAutogradHook wired into GradientTape's backward dispatcher. Backward replay touches LRU + rehydrates evicted streaming weights before each kernel reads them.

**Gap 4 — Fused dequant-matmul**: Q8MatMul + Q4MatMul AVX2/FMA paths with float-accum and per-tensor scale at group boundaries. int4 unpacks via PackedInt4 nibble accessors then reuses the int8 path.

**Gap 5 — Real native offload allocators (NOT placeholders)**:
* OpenCL: clCreateBuffer(CL_MEM_ALLOC_HOST_PTR) + clSVMAlloc, with full ICD context creation.
* Vulkan: vkCreateInstance + vkEnumeratePhysicalDevices + first-HOST_VISIBLE memory type → vkAllocateMemory + vkMapMemory.
* Metal: posix_memalign at 16K page alignment for MTLBuffer bytesNoCopy.
* WebGPU: wgpuCreateInstance + buffer pair.

**Gap 6 — Mixed-precision autodiff**: MixedPrecision.cs ships GradScaler with PyTorch-parity dynamic loss-scale schedule (init 65536, growth interval 2000, growth/backoff 2.0/0.5) + MasterWeights for fp32 master + low-precision compute pattern (Micikevicius et al. 2018).

**Gap 7 — QAT (quantization-aware training)**: QatTrainingHook with FakeQuantize round-trip on every forward read (STE backward), OptimizerStep, and ExportInt8/ExportInt4 inference payload (Jacob et al. 2018).

**Tests**: 32 new integration tests across all 7 gaps. Solution-wide build clean (13/13 csproj on net10.0 + net471). Full suite **4198/4198 passing — zero regressions**.
